### PR TITLE
[Compiler+VM PoC] Refactor instruction encoding/decoding

### PR DESCRIPTION
**Pull request by turbolent**
## Description

- Introduce an `Instruction` interface and define concrete types for all instructions
- Refactor the encoding functions to use the instruction type as an input (receiver)
- Refactor the decoding functions to output the instruction types
- Refactor the printing function into separate methods on the instruction types
- Refactor the compiler to emit instruction types
- Add a code gen that gathers the emitted instructions as-is, without encoding them to bytecode 

______

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: bbq/opcode/emit.go
Comment (SupunS) [inline bbq/opcode/emit.go:1]: Would it make sense to rename this file to something like `instructions.go`?

**Review by SupunS — APPROVED** [review]
👌

File: bbq/opcode/emit.go
Comment (turbolent) [reply bbq/opcode/emit.go:1]: Ah, yes, the name no longer fits, I'll rename it 👍

Comment (turbolent) [reply bbq/opcode/emit.go:1]: Done in [a0c6faf](https://github.com/onflow/cadence/pull/3714/commits/a0c6fafeed54150b762f19a7f169ca895e49c957)

**Pull request merged**